### DISCUSSION TITLE
backprop.BPTT function when using neuromorphic datasets

### DISCUSSION THREAD

**Discussion by msbouanane** [jeshraghian/snntorch#74] [Q&A]
Hello, 
The backprop.BPTT function requires the num_steps parameter. But when using a neuromorphic dataset, the num_steps will be variable depending on the longest recording in the mini-batch.
What should I pass to this parameter instead?

**Comment by jeshraghian** [comment]
The samples will have to be made even across the time dimension to make them usable in a DataLoader, and a DataLoader is required to perform batch training with `backprop.BPTT`.

If you are loading data from Tonic, then just pass `tonic.collation.PadTensors()` as an argument when you create the DataLoader, which automatically pads shorter samples to match the number of steps of the longest sample:

`trainloader = DataLoader(trainset, batch_size=batch_size, collate_fn=tonic.collation.PadTensors())`

Working example is provided in this tutorial / notebook: https://snntorch.readthedocs.io/en/latest/tutorials/tutorial_7.html

**Reply by msbouanane** [reply]
Thank you very much for your reply. I did create the DataLoader with `tonic.collation.PadTensor()` to make the samples even across the time dimension. However I think you didn't understand my point. In tutorial 7 (NEUROMORPHIC DATASETS WITH TONIC + SNNTORCH), the `snn.backprop` module (and hence the `backprop.BPTT` function) is NOT used to simplify the code. It is used in tutorial 6 section 3.3 (TRAINING AUTOMATION USING SNN.BACKPROP) with a static dataset. When I tried to use the `backprop.BPTT` function with a neuromorphic dataset to make the code easier, I specified the `time_var` argument to `True` but I didn't know what to do with the `num_steps` argument. In the documentation, it is mentioned that it must only be specified in case data in is static (https://snntorch.readthedocs.io/en/latest/_modules/snntorch/backprop.html#BPTT). When I removed it entirely, I got an error that an argument is missing. 
I hope you understand my question.

**Reply by jeshraghian** [reply]
Thanks for clarifying - this makes perfect sense now.
I'll go ahead and convert this discussion into an issue as you've raised an important point.

Different mini-batches have different lengths to each other. 
But `backprop.BPTT` requires a fixed time step. 

I'll modify the `backprop.BPTT` function so that `num_steps` is not required for neuromorphic data.
I'll try to handle it this week. The comment in the source code you're referring to must have been a note-to-self which fell through the cracks. 

For an immediately working 'hack', if you're using the N-MNIST dataset (as in tutorial 7), the final few time steps usually don't have much meaningful information. It is likely you can safely clip them, and simply use `num_steps` of the dataloader with the smallest number of time steps:

```
time_lengths = []

for data, target in iter(trainloader):
  time_lengths.append(data.size(0))

num_steps = min(time_lengths)

```

Then use `backprop.BPTT` in the following way:

`avg_loss = backprop.BPTT(net, trainloader, num_steps=num_steps, optimizer=optimizer, time_var=True, criterion=loss_fn)`

To avoid clipping, then you will need to write up the training loop (as in Tutorial 7), until I address this issue. Appreciate the heads up here.

**Reply by jeshraghian** [reply]
Just letting you know that I've just pushed v0.4.11 of snnTorch to PyPi / Conda which should hopefully fix this.

Additionally, one way to keep all batches of the same temporal duration is to specify the `n_time_bins` argument using Tonic's in-built transforms. I've done a sample of this in the following notebook:

https://colab.research.google.com/drive/1FWwpEyzf1Xx9xRDVnDUsjIxNCx_js2N4?usp=sharing

**Reply by msbouanane** [reply]
Thank you very much for you hacks and all of your help. And yes, the issue has been fixed. I have just trained my network with a neuromorphic dataset using `backprop.BPTT` and it seems to work perfectly.
